### ISSUE TITLE
Please support running a "first run" app

### ISSUE THREAD

**Issue by agx** [samcday/phrog#21]
When the user first boots the device (and no user exists yet) it would be nice to be able to (optionally) run a "first run" app fullscreen *before* the initial user selection is shown. This would allow to set up a default user / pin, etc. 

Phrog could spawn that app before it shows the `PhoshLockscreen`.

(The first run app could e.g. be gnome-initial-setup in it's system mode but which app it is likely doesn't matter for phrog).

I'm a bit fuzzy how we deal with permissions yet as we'd e.g. need polkit configuration that would allow the greetd user to create other users via accountsservice.

**Comment by samcday** [comment]
This sounds very interesting indeed. I imagine this would be a big help towards producing a generic image that can be "installed" (customized) on first boot. Aside from the `phrog` bits, I imagine there would be a lot of distro-specific work to achieve something like this.

What mechanism do you think would be best for this? I think it should probably just be some GSettings properties. If `com.samcday.phrog.firstrun-execute` is set with a string array for exec+args, then it boots directly into this instead of the lockscreen.

What if first-run process fails in some way? Should `phrog` be responsible for clearing the `firstrun-execute` prop when it has launched said process? Or should that responsibility lie elsewhere?

**Comment by agx** [comment]
> I imagine this would be a big help towards producing a generic image that can be "installed" (customized) on first boot.

Yes, that would be really nice. The current way is to run gnome-initial-setup after the first login but that requires that the user knows the initial PIN and usually can't customize the user name upfront.

> What mechanism do you think would be best for this? I

An initial naive approach could be: Run the "first run app" every time and let it decide whether it finished successfully (at the cost of maybe slightly slower startup). That would shift the responsibility out of phrog.

gnome-initial-setup used to work that way by exiting early when it found a file in the file system (it ran successfully already).

**Comment by samcday** [comment]
I'm gonna try hacking up an initial branch for this soon. I think it's gonna need some upstream Phosh changes.

We need a bit more control over the `home` widget of `PhoshShell`, I think. We ought to disable the default `PhoshHome` entirely - I can't think of any scenario where we want to present a user with a list of apps available to the greeter ;)

So in the case that a first-run app is configured, we don't lock on startup, and instead spawn the desired app. Once we detect the spawned app has exited we force the shell to be locked.

I'm a bit hazy on the specifics around this point. In particular, it'd be neat to hold the initial shell reveal until either a) the first-run app has loaded and signalled that its done its first paint b) a spinner tracking the app load has painted or c) a combination of the above? (wait for app first paint for N secs and then fall back to fading in the spinner)

**Comment by agx** [comment]
I see to ways:

1. Leave phosh out this 

like:

- if first run: launch configured app
- once first launch app exits: launch phosh

That would leverage the fact that we have control over when exactly the shell object is instantiated. We might need to unclutter `phosh_shell_constructed` for that at some point (so we can e.g. init CSS and make `PhoshFader` usable for fading without bringing up the panels. Downside is that we don't have the top panel for showing connection status.

2. Allow to not create `PhoshHome` (or at least put it behind the application)

> In particular, it'd be neat to hold the initial shell reveal 

If you have the full shell then you should have a splash screen doing this job.

**Comment by samcday** [comment]
> 1. Leave phosh out this

Mmm, I don't think that's a viable option. As you note, the top panel is important. Connection status and like, maybe changing device orientation and brightness (to actually conduct the First-run Experience :tm: properly), and so on.

> 2. Allow to not create PhoshHome (or at least put it behind the application)

Right, so I will start hacking some commits in the local `phosh/` subtree to this effect. Do you think it makes sense to dispatch the `phosh_home_new` call through a vtable func on PhoshShell (similar to how we virtualized `get_lockscreen_type`)?

> If you have the full shell then you should have a splash screen doing this job.

Well the issue is the fade-in signal to phoc happens at the end of the idle setup callback in PhoshShell. Currently that's fine because the lockscreen construction and populating the user+session happens synchronously during that init flow. If instead we launch a first-run app in a separate process, we're not blocking the PhoshShell constructor for that and so it's harder to ensure that initial fade-in happens in the sexiest way possible.

**Comment by agx** [comment]
> Mmm, I don't think that's a viable option

Yeah, that was my thinking as we'd not have connection status, BT devices, etc.

>  Do you think it makes sense to dispatch the phosh_home_new call through a vtable func on PhoshShell (similar to how we virtualized get_lockscreen_type)?

I'd rather have a property on `PhoshShell` that decides whether the current shell object has a `PhoshHome` rendered or not. That could become useful for other use cases like fullscreen.

**Comment by samcday** [comment]
> I'd rather have a property on PhoshShell that decides whether the current shell object has a PhoshHome rendered or not.

Cool, that makes sense. I'll do that :+1:

**Comment by samcday** [comment]
Very optimistically targeting the v0.42.0 release for this. Code freeze is 2 weeks away and I do not yet have any commits for consideration upstream yet.

**Comment by samcday** [comment]
I've opened https://gitlab.gnome.org/World/Phosh/phosh/-/merge_requests/1601 to get the ball rolling here upstream. It's a simple property on `Phosh.Shell` that prevents `PhoshHome` being constructed. I think the proposed patch is quite sane and unintrusive. We'll see what Guido thinks!

**Comment by samcday** [comment]
The initial stab is up as a PR - as such I think it's feasible to land this in 0.45

**Comment by samcday** [comment]
Upstream effort is hitting roadblocks, no longer targeting 0.45 for this work

**Comment by agx** [comment]
Just to be sure: As I noted in the [MR](https://gitlab.gnome.org/World/Phosh/phosh/-/merge_requests/1601#note_2312239) the  "API" for phrog remains unchanged so I wouldn't know how it changes anything on the phrog side.

**Comment by samcday** [comment]
> Just to be sure: As I noted in the [MR](https://gitlab.gnome.org/World/Phosh/phosh/-/merge_requests/1601#note_2312239) the "API" for phrog remains unchanged so I wouldn't know how it changes anything on the phrog side.

It's not about anything changing on the phrog side (the work on this side is fairly simple anyways), it's more about whether the necessary upstream changes will land in the 0.45 cycle at all ;)

**Comment by samcday** [comment]
Or just to put it slightly differently, I'm trying to target issues/PRs in this tracker against milestones that are mostly under my control. That is, not requiring movement on the libphosh functionality or API. The exception is when a proposed change has already had a thorough review upstream and has already received the `next` label or added to the corresponding upstream milestone.

**Comment by samcday** [comment]
The ability to toggle `PhoshHome` via the `overview-visible` property landed upstream, so we're totes on track for getting this feature out in 0.45 🤘

**Comment by samcday** [comment]
The initial support for this will be released in 0.45. The behaviour is fairly simple:

 * If `mobi.phosh.phrog.first-run` is set, it's spawned as a subprocess.
 * The shell will not be locked until this subprocess has exited.

This should make it fairly flexible to handle any kind of initial first-run wizard / flow. Phrog's shell doesn't have a `PhoshHome`, so there's no app list / search / window navigation. But the top panel is there of course so if someone boots up their new phone install at 1am and the first-run wizard doesn't have a dark theme they can still change brightness settings and prevent their retinas from being incinerated 🙃 

If there's any further behaviour/functionality desired, please open a follow issue!

**Comment by agx** [comment]
Will that setting be cleared if the first run app exists successfully?

**Comment by samcday** [comment]
Nope, I took the approach you touched on originally when you commented:

> Run the "first run app" every time and let it decide whether it finished successfully (at the cost of maybe slightly slower startup). That would shift the responsibility out of phrog.

So phrog doesn't manage the state of that setting at all, it just follows it. It'll be up to distros / first-run app to handle clearing the setting (or exiting early if first-run has already completed)

**Comment by agx** [comment]
👍 - that approach works perfectly well for me.

**Comment by agx** [comment]
Just in case someone comes across this and wants to work on a nice (e.g gtk-rs) app that allows to setup the system after first boot: we track that at https://gitlab.gnome.org/guidog/meta-phosh/-/issues/17 .
